tsdb/chunkenc/histogram_meta.go

```diff
@@ -359,92 +363,118 @@ loop:
-			// When deltas==false, it means that it is an absolute value. So we set it to 0 directly.
+		if ii < len(inserts) && i == inserts[ii].pos {
+			// We have an insert!
+			// Add insert.num new delta values such that their
+			// bucket values equate 0. When deltas==false, it means
+			// that it is an absolute value. So we set it to 0
+			// directly.
 			if deltas {
-				out[j] = -v
+				out[oi] = -v
 			} else {
-				out[j] = 0
+				out[oi] = 0
 			}
-			j++
-			for x := 1; x < interjections[interj].num; x++ {
-				out[j] = 0
-				j++
+			oi++
+			for x := 1; x < inserts[ii].num; x++ {
+				out[oi] = 0
+				oi++
 			}
-			interj++
+			ii++
 
 			// Now save the value from the input. The delta value we
 			// should save is the original delta value + the last
-			// value of the point before the interjection (to undo
-			// the delta that was introduced by the interjection).
-			// When deltas==false, it means that it is an absolute value,
+			// value of the point before the insert (to undo the
+			// delta that was introduced by the insert). When
+			// deltas==false, it means that it is an absolute value,
 			// so we set it directly to the value in the 'in' slice.
 			if deltas {
-				out[j] = d + v
+				out[oi] = d + v
 			} else {
-				out[j] = d
+				out[oi] = d
 			}
-			j++
+			oi++
 			v = d + v
 			continue
 		}
-
-		// If there was no interjection, the original delta is still
-		// valid.
-		out[j] = d
-		j++
+		// If there was no insert, the original delta is still valid.
+		out[oi] = d
+		oi++
 		v += d
 	}
-	switch interj {
-	case len(interjections):
-		// All interjections processed. Nothing more to do.
-	case len(interjections) - 1:
-		// One more interjection to process at the end.
+	switch ii {
+	case len(inserts):
+		// All inserts processed. Nothing more to do.
+	case len(inserts) - 1:
+		// One more insert to process at the end.
 		if deltas {
-			out[j] = -v
+			out[oi] = -v
 		} else {
-			out[j] = 0
+			out[oi] = 0
 		}
-		j++
-		for x := 1; x < interjections[interj].num; x++ {
-			out[j] = 0
-			j++
+		oi++
+		for x := 1; x < inserts[ii].num; x++ {
+			out[oi] = 0
+			oi++
 		}
 	default:
-		panic("unprocessed interjections left")
+		panic("unprocessed inserts left")
 	}
 	return out
 }
+
+// counterResetHint returns a CounterResetHint based on the CounterResetHeader
+// and on the position into the chunk.
+func counterResetHint(crh CounterResetHeader, numRead uint16) histogram.CounterResetHint {
+	switch {
+	case crh == GaugeType:
+		// A gauge histogram chunk only contains gauge histograms.
+		return histogram.GaugeType
+	case numRead > 1:
+		// In a counter histogram chunk, there will not be any counter
+		// resets after the first histogram.
+		return histogram.NotCounterReset
+	case crh == CounterReset:
+		// If the chunk was started because of a counter reset, we can
+		// safely return that hint. This histogram always has to be
+		// treated as a counter reset.
+		return histogram.CounterReset
+	default:
```

Comment:
I think in this case we should take the counter reset hint from the chunk's meta. If the user of this chunk (for example the Head block or the persistent block) sets it right, it will be useful to know the right hint for the first sample.

Reply:
Yes, I do believe there are scenarios where we can indeed return `histogram.NotCounterReset` (mostly if we know this chunk is not the first chunk in a block), but this needs further caution higher up the stack, which this PR doesn't include. I mostly want to be conservative for now. We can later evaluate if the additional work (and risk) is worth it.
Note that the case where a chunk is created with `NotCounterReset` is quite rare. The only reason that comes to mind right now is if the previous chunk is full, which currently means 120 histograms in the previous chunks.  Since a new chunk will often be cut because of a counter reset, there are even fewer than one in 120 histograms that will run into the `NotCounterReset` case, so the performance gains will be quite small.

I will add a TODO here anyway to not forget to make this trade-off later.